keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,672 @@
+  // - cloud.provider.com/topology-rack
+	NodeLabel string
+}
+```
+
+Example TAS Resource Flavor & Topology configuration:
+
+```yaml
+kind: ResourceFlavor
+metadata:
+  name: "tas-flavor"
+spec:
+  nodeLabels:
+    "cloud.provider.com/node-group: tas"
+  topologyName: default
+---
+kind: Topology
+metadata:
+  name: "default"
+spec:
+  levels:
+  - nodeLabel: cloud.provider.com/topology-block
+  - nodeLabel: cloud.provider.com/topology-rack
+```
+
+### User-facing API
+
+The user will need to point the workload to the ClusterQueue with the TAS
+ResourceFlavor, and add one of the annotations:
+
+```golang
+const (
+
+  // This annotation indicates that a workload requires Topology Aware Scheduling,
+  // and running all pods on nodes closely connected within the same level of
+  // hierarchy is a strong requirement for scheduling the workload.
+  RequireTopologyAnnotation = "kueue.x-k8s.io/require-topology"
+
+  // This annotation indicates that a workload requires Topology Aware Scheduling,
+  // but running all pods without the same topology level is a preference rather
+  // than requirement. There is a distinguished value "auto" which means that
+  // the lowest hierarchy level should be used.
+  PreferTopologyAnnotation = "kueue.x-k8s.io/prefer-topology"
+)
+```
+
+### Validation
+
+The validations at the creation time:
+- the ResourceFlavor defining`TopologyAwareScheduling` needs to have at least
+  one node label
+
+We introduce the following validations (post-creation time, workload violating
+the rules is deactivated):
+- the value of `kueue.x-k8s.io/require-topology` is one of the labels specified
+  in the topology node labels
+- the value of `kueue.x-k8s.io/prefer-topology` is one of the labels specified
+  in the topology node labels
+
+### Internal APIs
+
+We extend the `Workload` structure to reflect the topology request at the
+Job level.
+
+```golang
+type WorkloadSpec struct {
+  ...
+  // TopologyRequest defines the topology requested for the corresponding Job.
+  TopologyRequest *TopologyRequest
+}
+
+type TopologyRequest struct {
+  // Policy defines the policy used for TAS. Possible values are:
+  // - Prefer set when `kueue.x-k8s.io/prefer-topology` annotation is set on the Job
+  // - Require set when `kueue.x-k8s.io/require-topology` annotation is set on the Job
+  Policy TopologyRequestPolicy
+
+  // Level indicated by the `kueue.x-k8s.io/prefer-topology` or `kueue.x-k8s.io/require-topology `
+  // annotation
+  Level string
+}
+```
+
+We extend the `PodSetAssignment` structure to keep track of the number of pods
+at each topology level to the specific subset of nodes.
+
+```golang
+type PodSetAssignment struct {
+  ...
+
+  // TopologyAssignment indicates the resources assigned per topology level
+  TopologyAssignment *TopologyAssignment
+}
+
+type TopologyAssignment struct {
+     // Slices contains the list of assignments split into slices
+     Slices []TopologyAssignmentSlice
```

Comment:
AFAIK, there are no "slices" concept in the k/k world. So, I would like to rephrase this concept to align with the k/k world.

Hence, could you explain the motivation why you would like to put this dedicated type instead of putting nodeLabels and found here?

Reply:
Agree, slices can be confusing as it is used at various contexts meaning slightly different things. Maybe "Fragments" instead of slices?

Reply:
Updated to `TopologyAssignmentGroup`. PTAL.